### PR TITLE
Update 2021 eng org plan

### PR DESCRIPTION
**Pull request by nicksnyder**
This updates our documentation with a new forecast for our 2021 engineering org.

Inputs to this document were:
- [My thinking/notes on org structure](https://workflowy.com/s/eng-org-chart/wSkGQstPvRZXtFZ5)
- Our WIP [2021 headcount plan for engineering](https://docs.google.com/spreadsheets/d/1UG4aLB5S8yOSOByia2EQgpZkoPxG5eVfqQqgs1f71Vw/edit#gid=2097845084)
- WIP [Web team growth plan](https://docs.google.com/document/d/1lzlsu9SoGSou1PTK7gp75ueYLKYitRtB8t8TxBk1irY/edit#heading=h.s8kaxupckw0l)
- Other existing growth plans already documented in the handbook (I edited formatting here and there).
- [Team survey on "infrastructure" vs "platform" terminology](https://sourcegraph.slack.com/archives/C07KZF47K/p1605656483057100)

### CODE REVIEW

**Review by christinaforney — APPROVED** [review]
I like this!

File: handbook/engineering/2021_org.md [hunk 1]
```diff
@@ -1,46 +1,32 @@
-# 2021 Q1 Organizational Plan
+# 2021 Engineering Organization Plan
 
 Defining a future vision for the organization is necessary to effectively grow the engineering team. It is understood that all plans change, and this document will be kept up to date with these changes. This is our written plan to align the team and clarify hiring and growth expectations.
 
+[2021 headcount plan](https://docs.google.com/spreadsheets/d/1UG4aLB5S8yOSOByia2EQgpZkoPxG5eVfqQqgs1f71Vw/edit#gid=2097845084)
```
Comment (ryanslade) [inline handbook/engineering/2021_org.md:5]: Are the years at the top of the spreadsheet correct? The earliest is Q3 2021 which is a a year away :)

Comment (ryanslade) [reply handbook/engineering/2021_org.md:5]: I guess it's financial years in which case it makes sense :)

**Review by mrnugget — COMMENTED** [review]
I have a question regarding the split into "Product" and "Infrastructure"/"Platform" with the latter including "Backend infra/platform" and "Dev tooling and happiness" [in your notes](https://workflowy.com/s/eng-org-chart/wSkGQstPvRZXtFZ5):

Do you still think we should have a separate "Dev tooling and happiness" team? Or does that fall under "Backend Platform" in this PR?

In both cases, much much more so if "Dev tooling" is a separate team: what are your thoughts on ensuring that these teams have skin in the game, and dogfood their own tooling?

I'm asking because our dev environment is already growing and getting slower and more unwieldy. One of the reasons (besides: we're growing! our product is growing!) is that it's easy for everyone to _add_ things to the environment (to mirror our production/customer envs). The problem is that if you don't use the dev env daily (booting it up, restarting, waiting for reloads) you don't notice how it affects the daily developer experience and don't think twice about adding things.

In other words: how do you ensure that separate dev-tooling/infrastructure teams are solving the pains that are holding the product teams back vs. building the things they _suspect_ would be good to have?

**Comment by chayim** [comment]
> * Is it intuitively clear to you what the boundaries are between Cloud, Deployment and Operations Platform, Developer Platform, Backend Platform and Web Platform?

I think we should remove the term Platform in general - though specifically I think it makes perfect sense for Developer Platform.

**Review by chayim — APPROVED** [review]
This looks great!

File: handbook/engineering/2021_org.md [hunk 2]
```diff
@@ -1,46 +1,32 @@
...
+    - [Extensions and Integrations](web/index.md#extensions-and-integrations) (future team, currently owned by Web team)
+  - Director of Engineering - Infrastructure
+    - [Cloud](cloud/index.md)
+    - [Deployment and Operations Platform](distribution/index.md#deployment-and-operations-platform)
+    - [Developer Platform](distribution/index.md#developer-platform)
+    - [Security](security/index.md)
+    - [Backend Platform](backend-infrastructure/index.md) (future team, currently owned by Cloud team)
+    - [Web Platform](web/index.md#web-infrastructure) (future team, currently owned by Web team)
+
+Every team here is a "full stack" team that has all the skills and roles necessary to achieve the team's mission.
```
Comment (lguychard) [inline handbook/engineering/2021_org.md:23]:
```suggestion
Every team here is cross-functional, with all the skills and roles necessary to achieve the team's mission.
```

**Review by lguychard — APPROVED** [review]

File: handbook/engineering/2021_org.md (hunk 2 above)
Comment (mrnugget) [reply handbook/engineering/2021_org.md:23]: FYI, meta-comment: "cross-functional" always to me "across the _horizontal_ organisational structure": backend engineers, frontend engineers, _designers_, tech writers, etc. "full stack" makes me think vertically: from the browser down to the database/deployment.

Comment (lguychard) [reply handbook/engineering/2021_org.md:23]: I have the same perception, and since we highlight "skills and roles" in the same sentence, cross-functional makes more sense to me here.

**Comment by tsenart** [comment]
> In other words: how do you ensure that separate dev-tooling/infrastructure teams are solving the pains that are holding the product teams back vs. building the things they suspect would be good to have?

My take on this is that platform / infrastructure teams need to listen to their internal customers, just like Sourcegraph has to listen to their external customers. Roadmaps for those teams need to be designed and prioritised based on a mix of real current pains and future pains. Those teams will work on the right things by being customer focused AND data informed (i.e. quantify the pains accurately, and show them changing over time).

File: handbook/engineering/2021_org.md [hunk 3]
```diff
@@ -1,46 +1,32 @@
...
+  - Director of Engineering - Product
+    - [Search Core](search/index.md#search-core) (future split of current Search team)
+    - [Search Growth](search/index.md#search-growth) (future split of current Search team)
+    - [Campaigns](campaigns/index.md)
+    - [Code Intelligence](code-intelligence/index.md)
+    - [Code Insights](web/index.md#code-insights) (future team, currently owned by Web team)
+    - [Extensions and Integrations](web/index.md#extensions-and-integrations) (future team, currently owned by Web team)
+  - Director of Engineering - Infrastructure
+    - [Cloud](cloud/index.md)
+    - [Deployment and Operations Platform](distribution/index.md#deployment-and-operations-platform)
```
Comment (tsenart) [inline handbook/engineering/2021_org.md:17]: Not sure if better, but the previous name is very long.

```suggestion
    - [Production Platform](distribution/index.md#deployment-and-operations-platform)
```

**Review by tsenart — APPROVED** [review]
> I am considering using the word "Platform" instead of "Infrastructure". The point is to make it clear that these teams don't merely own a bunch of shared pieces, but they own the overall developer experience of the respective areas. Does this make sense to you?

I don't have the sense that the term infrastructure implies owning just shared pieces and not caring about the end to end internal customer needs, and I have a slight preference for that term, but I'm not opposed to platform.

**Comment by chrispine** [comment]
> how do you ensure that separate dev-tooling/infrastructure teams are solving the pains that are holding the product teams back vs. building the things they _suspect_ would be good to have?

@mrnugget Adding to what Tomás said, I think in the specific case of spin-up/shut-down times, these need to be explicitly tracked metrics for whichever team is responsible for this, and tracked over time, and a commitment to keeping those times down. That's how we solve this problem "once and for all!"

**Review by chrispine — APPROVED** [review]
LGTM

**Review by jeanduplessis — APPROVED** [review]

File: handbook/engineering/distribution/index.md [hunk 4]
```diff
@@ -56,8 +56,18 @@ Other:
...
+_Updated 2020-11-11_
 
-The team has doubled in size recently so it isn't a high priority to grow this team further, but we are always open to hiring exceptional people. [Apply here](https://github.com/sourcegraph/careers/blob/master/job-descriptions/software-engineer-distribution.md).
+We are looking to grow this team to 8 engineers so we can split this team into two new teams.
+
+TODO(@gonza): define specific roles we are hiring for.
+
+### Deployment and Operations platform
+
+This team owns how Sourcegraph site admins (including on-premise customers and our own Cloud deployment) deploy, upgrade, and monitor Sourcegraph.
```
Comment (pecigonzalo) [inline handbook/engineering/distribution/index.md:69]:
```suggestion
This team owns how Sourcegraph site admins (including on-premise customers and our own Cloud deployment) deploy, upgrade, operate and monitor Sourcegraph.
```

File: handbook/engineering/distribution/index.md [hunk 5]
```diff
@@ -56,8 +56,18 @@ Other:
...
+
+### Developer platform
+
+This team owns general developer productivity and happiness at Sourcegraph. The customers of this team are other engineers at Sourcegraph. Ownership areas include, but aren't limited to, our dev environment, our build pipelines, and developer bots/automation.
```
Comment (pecigonzalo) [inline handbook/engineering/distribution/index.md:73]:
```suggestion
This team owns general developer productivity and happiness at Sourcegraph. The customers of this team are other engineers at Sourcegraph. Ownership areas include, but aren't limited to, our dev tools, our build infrastructure, and developer bots/automation.
```
Im not entirely set on this change, but I think it would be good to expect teams to create and maintain the own chunks of the overall pipeline and build/dev workflow, as a central team can create the tools for generally automating and making processes easier (buildkite plugins, the release cli, "default" build pipelines for Go or other langs, etc) but each team has the domain specific knowledge around their systems.

File: handbook/engineering/2021_org.md (hunk 3 above)
Comment (pecigonzalo) [reply handbook/engineering/2021_org.md:17]: I think Distribution does better capture what is described as the objective and scope of the team. I dont think its a deal-breaker and we can rename in the future.

**Review by pecigonzalo — APPROVED** [review]

File: handbook/engineering/2021_org.md [hunk 6]
```diff
@@ -1,46 +1,32 @@
...
+Every team here is cross-functional, with all the skills and roles necessary to achieve the team's mission.
+
+Examples:
+
+- Product teams own their own on-call and operations, using infrastructure provided by Deployment and Operations Platform team.
+  - Is search broken/down? That is the search team's problem to solve.
+  - Is our whole Cloud deployment broken/down? That is the Cloud team's problem to solve.
+  - Is product team X lacking the tools they need to appropriately monitor their services? That is the Deployment and Operations Platform team's problem to solve.
+- Product teams own the security of their parts of the application. They should engage and consult with security experts on our team to get ahead of security concerns.
+- To a first approximation, each team, including infrastructure teams, need product managers. Each team has customers (for infrastructure teams, customers are internal). Most teams need designers too.
```
Comment (felixfbecker) [inline handbook/engineering/2021_org.md:32]: Does this mean each designer will be considered **members of** one certain team like each PM, as opposed to only being _primarily associated_ with a certain team?

File: handbook/engineering/web/index.md [hunk 7]
```diff
@@ -100,10 +100,18 @@ Before web team syncs, teammates and stakeholders should write down under "Discu
...
+
+The **Sourcegraph web application and codebase** is clean, cohesive, stable and performant. It's easy for all teammates to onboard and contribute to the web application.
+
+### Code insights
+
+Sourcegraph surfaces high-level [**code insights**](https://docs.google.com/document/d/1EHzor6I1GhVVIpl70mH-c10b1tNEl_p1xRMJ9qHQfoc/edit) to engineering leaders, empowering data-driven decisions.
+
+### Extensions and integrations
+
+**Sourcegraph extensions** empower users to integrate Sourcegraph with any third-party service providing useful information about code (code coverage, exception tracking, tracing, code quality). They are consistently supported across all code host integrations and the Sourcegraph UI.
```
Comment (felixfbecker) [inline handbook/engineering/web/index.md:115]: including, but not limited to
```suggestion
**Sourcegraph extensions** empower users to integrate Sourcegraph with any third-party service providing useful information about code (e.g. code coverage, exception tracking, tracing, code quality). They are consistently supported across all code host integrations and the Sourcegraph UI.
```

File: handbook/engineering/web/index.md [hunk 8]
```diff
@@ -100,10 +100,18 @@ Before web team syncs, teammates and stakeholders should write down under "Discu
...
 
-- Web infrastructure
-  - The **Sourcegraph web application and codebase** is clean, cohesive, stable and performant. It's easy for all teammates to onboard and contribute to the web application.
-- Extensions and integrations
-  - **Sourcegraph extensions** empower users to integrate Sourcegraph with any third-party service providing useful information about code (code coverage, exception tracking, tracing, code quality). They are consistently supported across all code host integrations and the Sourcegraph UI. Through extensions, Sourcegraph surfaces high-level [**code insights**](https://docs.google.com/document/d/1EHzor6I1GhVVIpl70mH-c10b1tNEl_p1xRMJ9qHQfoc/edit) to engineering leaders, empowering data-driven decisions.
-  - The **browser extension** and code host **native integrations** are a breeze to set up, and add compelling value when reading or reviewing code. Enabling native code host integrations for all users is a no-brainer for site admins.
+### Web infrastructure
```
Comment (felixfbecker) [inline handbook/engineering/web/index.md:105]:
```suggestion
### Web platform
```

File: handbook/engineering/2021_org.md [hunk 9]
```diff
@@ -1,46 +1,32 @@
...
+    - [Code Intelligence](code-intelligence/index.md)
+    - [Code Insights](web/index.md#code-insights) (future team, currently owned by Web team)
+    - [Extensions and Integrations](web/index.md#extensions-and-integrations) (future team, currently owned by Web team)
+  - Director of Engineering - Infrastructure
+    - [Cloud](cloud/index.md)
+    - [Deployment and Operations Platform](distribution/index.md#deployment-and-operations-platform)
+    - [Developer Platform](distribution/index.md#developer-platform)
+    - [Security](security/index.md)
+    - [Backend Platform](backend-infrastructure/index.md) (future team, currently owned by Cloud team)
+    - [Web Platform](web/index.md#web-infrastructure) (future team, currently owned by Web team)
```
Comment (felixfbecker) [inline handbook/engineering/2021_org.md:21]:
```suggestion
    - [Web Platform](web/index.md#web-platform) (future team, currently owned by Web team)
```

**Comment by felixfbecker** [comment]
> I am considering using the word "Platform" instead of "Infrastructure". The point is to make it clear that these teams don't merely own a bunch of shared pieces, but they own the overall developer experience of the respective areas. Does this make sense to you?

I much much prefer "platform" over "infrastructure" at least for the web team 👍  "Platform" implies that team owns the general UX too (e.g. navigation), while "infrastructure" sounds like they would only own shared component libraries.

File: handbook/engineering/2021_org.md (hunk 6 above)
Comment (nicksnyder) [reply handbook/engineering/2021_org.md:32]: I am not sure what the meaningful distinction is between those two options.

PMs and Designers are part of the functional team (but their manager is not the engineering manager).

In an ideal world, PMs and Designers are 1:1 with teams. In our current, and less than ideal world, we don't have enough PMs/designers so each PM/Designer might be involved with multiple teams.

File: handbook/engineering/2021_org.md (hunk 3 above)
Comment (nicksnyder) [reply handbook/engineering/2021_org.md]: Thanks for the feedback, I changed this.

File: handbook/engineering/distribution/index.md (hunk 5 above)
Comment (nicksnyder) [reply handbook/engineering/distribution/index.md]: I updated how I was thinking about this.

**Comment by nicksnyder** [comment]
I just pushed a non-trivial update to this plan based on questions and feedback I received, so please re-review.

Notable changes:

- Standardizing on "Platform" instead of "Infrastructure" because I think that communicates the intent better across the team (given survey I did in Slack). The only risk here is if we decide to build/treat Sourcegraph (the product) as a platform (e.g. our extension API) and that team also ends up with "platform" in its name
- In the steady state, I don't think there is a "Cloud" team as it exists today. In the proposed team structure and effort like Cloud would be lead by a PM to prioritize work that needs to be done by multiple engineering teams. Even in this new model, 80% of the effort would land on the backend platform team (which is what the Cloud team was before it became the Cloud team). When Cloud has achieved its goals, I think it makes sense to rebrand it back to the backend platform team (and frontend devs can go to web platform team). Ongoing operational support for Cloud will be handled by the proposed "Operations platform" team.

Thought experiment:

How would we prioritize adoption of Sourcegraph Cloud with the new team structure since there isn't a clear team that would own this? My answer: This would be a product initiative that would be lead by a PM who would be accountable for the outcome. That PM would decide how to best achieve that outcome. They might conclude that we need to increase the number of repositories on Sourcegraph.com, or that we need to do more partnerships with orgs like CNCF. Once the work is identified, they would identify what is the closest existing team (or teams) that they would need to do this work and then work with those teams/PMs to prioritize it. For bigger and more durable missions, we might decide to spin up a long running team.

> how do you ensure that separate dev-tooling/infrastructure teams are solving the pains that are holding the product teams back vs. building the things they suspect would be good to have?

I agree with Tomás and Chris. We solve this in the exact same way that we ensure our product teams are building the right things for our customers: by soliciting feedback from other engineers and building what they are asking for. I want to also point out that the existence of proposed platform teams doesn't preclude other teams from contributing here.

File: handbook/engineering/web/index.md [hunk 10]
```diff
@@ -91,10 +91,36 @@ Before web team syncs, teammates and stakeholders should write down under "Discu
...
+This team is responsible for how we bring the value of Sourcegraph to other developer tools as well as how we bring the value of other developers tools into Sourcegraph.
+
+Examples:
+
+- Surface code intelligence (and other Sourcegraph data) in code hosts through user installed browser extensions.
+- Add native support for Sourcegraph in code hosts (for example: GitLab) or popular sites where developer look at code examples (e.g. https://reactjs.org, https://pkg.go.dev/) so users don't have to install our browser extensions to get code intelligence.
+- Build an extension API for Sourcegraph that enables developers to bring data from their favorite dev tools (for example: code coverage, exception tracking, tracing, code quality) into their Sourcegraph workflow.
+- Build useful Sourcegraph extensions on top of the Sourcegraph extension API (for example: Codecov, Datadog, Sentry, Lightstep)
+
+This team will not be created until we have identified a dedicated engineering manager and a dedicated product manager. In the meantime
```
Comment (LawnGnome) [inline handbook/engineering/web/index.md:126]: There's either a missing end to this sentence, or this is an _excellent_ [Spacehog reference](https://www.youtube.com/watch?v=PCsGRCf8T9Y).

Comment (nicksnyder) [reply handbook/engineering/web/index.md:126]: Oops!

```suggestion
This team will not be created until we have identified a dedicated engineering manager and a dedicated product manager.
```

**Review by tsenart — COMMENTED** [review]
I'm almost comfortable with the latest revision of this, apart from one thing: Who owns the site admin experience (i.e. `/site-admin`)? This is a mix of backend and frontend, and possibly, deployment? Historically this has been the realm of the Core Services then Cloud team, but we were always understaffed on frontend skills. With Backend Platform, the focus of the team is tightened and it seems like the site admin area needs another owner?

**Comment by jeanduplessis** [comment]
> I'm almost comfortable with the latest revision of this, apart from one thing: Who owns the site admin experience (i.e. `/site-admin`)? This is a mix of backend and frontend, and possibly, deployment? Historically this has been the realm of the Core Services then Cloud team, but we were always understaffed on frontend skills. With Backend Platform, the focus of the team is tightened and it seems like the site admin area needs another owner?

@tsenart I think going forward this will likely be the Web Platform team. Each product team will be responsible for their functionality in the backend and the Web Platform team would support the frontend.

From the Web Platform team's responsibilities
> Product improvements that affect our entire app and don't have a clear product team owner (for example: new nav bar, complete site redesign)

File: handbook/engineering/2021_org.md [hunk 11]
```diff
@@ -1,46 +1,38 @@
...
-      - Backend Engineers: 3
-      - Security Engineers: 3
-    - [Web Infrastructure Team](web/index.md)
-      - Engineering Manager: Web Infrastructure
-      - Frontend Engineers: 4
+[2021 headcount plan](https://docs.google.com/spreadsheets/d/1UG4aLB5S8yOSOByia2EQgpZkoPxG5eVfqQqgs1f71Vw/edit#gid=2097845084)
+
+## Product engineering team
+
+This organization is responsible for building products that deliver value to our customers and drive business growth. Each product engineering team has the autonomy and the necessary cross-functional skills (for example: frontend engineers, backend engineers, site reliability engineers, product manager, designer) to achieve their mission.
```
Comment (jeanduplessis) [inline handbook/engineering/2021_org.md:9]: @nicksnyder do you intend for the site reliability engineering role to fall under the span of influence of the engineering manager of the product team? Or will they be a counterpart to the team like the PM and Designer is and report to a different manager?

**Review by lguychard — APPROVED** [review]
LGTM!

File: handbook/engineering/2021_org.md (hunk 11 above)
Comment (nicksnyder) [reply handbook/engineering/2021_org.md:9]: Good question. My default instinct is to follow general principles (teams should have full stack ownership) and say necessary SREs will be embedded on the team and report to the EM of the product team. If our monitoring/alerting infra is good enough (built by operations platform team), then it might be unnecessary to have specialized SREs on product teams (frontend/backend engineers can manage the alerts for those areas of the product).

File: handbook/engineering/2021_org.md [hunk 12]
```diff
@@ -1,46 +1,32 @@
...
-      - Frontend Engineers: 4
+  - Director of Engineering - Product
+    - [Search Core](search/index.md#search-core) (future split of current Search team)
+    - [Search Growth](search/index.md#search-growth) (future split of current Search team)
+    - [Campaigns](campaigns/index.md)
+    - [Code Intelligence](code-intelligence/index.md)
+    - [Code Insights](web/index.md#code-insights) (future team, currently owned by Web team)
+    - [Extensions and Integrations](web/index.md#extensions-and-integrations) (future team, currently owned by Web team)
+  - Director of Engineering - Infrastructure
+    - [Cloud](cloud/index.md)
```
[the remark this answers is not in the document]
Comment (nicksnyder) [reply handbook/engineering/2021_org.md]: Yeah, this is confusing. The "Cloud" effort is a product effort that requires mostly, but not all, backend work. In the steady state, Christina does not want to treat "Cloud" as a separate product, it really isn't a permanent team at all I think (which is what my latest update reflects).

**Review by jeanduplessis — APPROVED** [review]

**Comment by nicksnyder** [comment]
This plan isn't perfect, and there are still some open questions, but I think it is better than our currently documented plan, so I am going to merge now. We will continue to adapt this plan over time as we learn new information.

As a next step, eng managers are doing an async process to stress test this org chart against all the possible things we could possibly do in 2021. WIP https://docs.google.com/document/d/1YcGzp2deYAMZnuGGe2hnA1Lc02polkNv0QVyu8Rm6jg/edit
